File: sdks/python/roost/resources/installer_deployments.py
"""``roost.installer_deployments`` — classic-installer deployment lifecycle.

Drives the wave-1A site-scoped routes:

  POST   /api/sites/{siteId}/deployments
  GET    /api/sites/{siteId}/deployments
  GET    /api/sites/{siteId}/deployments/{deploymentId}
  POST   /api/sites/{siteId}/deployments/{deploymentId}/retry
  POST   /api/sites/{siteId}/deployments/{deploymentId}/cancel
  POST   /api/sites/{siteId}/deployments/{deploymentId}/uninstall

Note: this is the *classic installer* deploy noun, distinct from the
content-addressed atomic roost deploy (``roost.roosts.deploy``). The
two surfaces share no state.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from roost.client import RoostClient


@dataclass(slots=True)
class InstallerDeploymentTarget:
    machine_id: str
    status: str
    error: str | None = None


@dataclass(slots=True)
class InstallerDeploymentSummary:
    id: str
    name: str
    installer_name: str
    installer_url: str
    silent_flags: str
    verify_path: str | None
    sha256_checksum: str | None
    parallel_install: bool
    targets: list[InstallerDeploymentTarget] = field(default_factory=list)
    status: str = "pending"
    created_at: str | None = None
    completed_at: str | None = None
    updated_at: str | None = None


@dataclass(slots=True)
class InstallerDeploymentDetail:
    id: str
    site_id: str
    name: str
    installer_name: str
    installer_url: str
    silent_flags: str
    verify_path: str | None
    sha256_checksum: str | None
    parallel_install: bool
    targets: list[InstallerDeploymentTarget] = field(default_factory=list)
    status: str = "pending"
    created_at: str | None = None
    completed_at: str | None = None
    updated_at: str | None = None


def _parse_target(raw: dict[str, Any]) -> InstallerDeploymentTarget:
    err = raw.get("error")
    return InstallerDeploymentTarget(
        machine_id=str(raw.get("machineId", "")),
        status=str(raw.get("status", "")),
        error=str(err) if isinstance(err, str) else None,
    )


def _parse_summary(raw: dict[str, Any]) -> InstallerDeploymentSummary:
    targets = [
        _parse_target(t)
        for t in (raw.get("targets") or [])
        if isinstance(t, dict)
    ]
    return InstallerDeploymentSummary(
        id=str(raw.get("id", "")),
        name=str(raw.get("name", "")),
        installer_name=str(raw.get("installer_name", "")),
        installer_url=str(raw.get("installer_url", "")),
        silent_flags=str(raw.get("silent_flags", "")),
        verify_path=raw.get("verify_path"),
        sha256_checksum=raw.get("sha256_checksum"),
        parallel_install=bool(raw.get("parallel_install", False)),
        targets=targets,
        status=str(raw.get("status", "pending")),
        created_at=raw.get("createdAt"),
        completed_at=raw.get("completedAt"),
        updated_at=raw.get("updatedAt"),
    )


def _parse_detail(raw: dict[str, Any]) -> InstallerDeploymentDetail:
    targets = [
        _parse_target(t)
        for t in (raw.get("targets") or [])
        if isinstance(t, dict)
    ]
    return InstallerDeploymentDetail(
        id=str(raw.get("id", "")),
        site_id=str(raw.get("siteId", "")),
        name=str(raw.get("name", "")),
        installer_name=str(raw.get("installer_name", "")),
        installer_url=str(raw.get("installer_url", "")),
        silent_flags=str(raw.get("silent_flags", "")),
        verify_path=raw.get("verify_path"),
        sha256_checksum=raw.get("sha256_checksum"),
        parallel_install=bool(raw.get("parallel_install", False)),
        targets=targets,
        status=str(raw.get("status", "pending")),
        created_at=raw.get("createdAt"),
        completed_at=raw.get("completedAt"),
        updated_at=raw.get("updatedAt"),
    )


class InstallerDeployments:
    """Classic agent-installer deployments (wave 1A)."""

    def __init__(self, client: "RoostClient") -> None:
        self._client = client

    async def list(
        self,
        site_id: str,
        *,
        page_size: int | None = None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        """List deployments. Returns the raw page envelope ``{items, next_page_token}``."""
        query: dict[str, Any] = {}
        if page_size is not None:
            query["page_size"] = page_size
        if page_token:
            query["page_token"] = page_token
        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments",
            query=query or None,
        )
        data = resp.data if isinstance(resp.data, dict) else {}
        items = [
            _parse_summary(item)
            for item in (data.get("items") or [])
            if isinstance(item, dict)
        ]
        return {
            "items": items,
            "next_page_token": str(data.get("next_page_token") or ""),
        }

    async def get(self, site_id: str, deployment_id: str) -> InstallerDeploymentDetail:
        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments/{deployment_id}",
        )
        return _parse_detail(resp.data if isinstance(resp.data, dict) else {})

    async def create(
        self,
        site_id: str,
        *,
        name: str,
        installer_name: str,
        installer_url: str,
        silent_flags: str,
        machines: Sequence[str],
        verify_path: str | None = None,
        sha256_checksum: str | None = None,
        parallel_install: bool = False,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "name": name,
            "installer_name": installer_name,
            "installer_url": installer_url,
            "silent_flags": silent_flags,
            "machines": list(machines),
        }
        if verify_path is not None:
            body["verify_path"] = verify_path
        if sha256_checksum is not None:
            body["sha256_checksum"] = sha256_checksum
        if parallel_install:
            body["parallel_install"] = True

        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments",
            method="POST",
            body=body,
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def retry(
        self,
        site_id: str,
        deployment_id: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments/{deployment_id}/retry",
            method="POST",
            body={},
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def cancel(
        self,
        site_id: str,
        deployment_id: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments/{deployment_id}/cancel",
            method="POST",
            body={},
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def uninstall(
        self,
        site_id: str,
        deployment_id: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/sites/{site_id}/deployments/{deployment_id}/uninstall",
            method="POST",
            body={},
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}


__all__ = [
    "InstallerDeploymentDetail",
    "InstallerDeploymentSummary",
    "InstallerDeploymentTarget",
    "InstallerDeployments",
]

File: sdks/node/src/resources/installerDeployments.ts
/**
 * `roost.installerDeployments` - classic installer deployment lifecycle.
 *
 * Wraps the wave-1A installer-deployments routes:
 *
 *   GET    /api/sites/{siteId}/deployments
 *   POST   /api/sites/{siteId}/deployments
 *   GET    /api/sites/{siteId}/deployments/{deploymentId}
 *   POST   /api/sites/{siteId}/deployments/{deploymentId}/retry
 *   POST   /api/sites/{siteId}/deployments/{deploymentId}/cancel
 *   POST   /api/sites/{siteId}/deployments/{deploymentId}/uninstall
 *   DELETE /api/sites/{siteId}/deployments/{deploymentId}
 *
 * Every mutation auto-generates an `Idempotency-Key` of the form
 * `sdk-installer-deployments-<verb>-${randomUUID()}` if the caller does
 * not pass `opts.idempotencyKey`. Errors surface as the canonical
 * `RoostApiError` instances thrown by `client.request`.
 */
import { randomUUID } from 'crypto';
import type { RoostClient } from '../lib/client';

/* --------------------------------------------------------------------- */
/*  types                                                                */
/* --------------------------------------------------------------------- */

export type InstallerDeploymentTargetStatus =
  | 'pending'
  | 'closing_processes'
  | 'downloading'
  | 'installing'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'uninstalled';

export type InstallerDeploymentStatus =
  | 'pending'
  | 'in_progress'
  | 'completed'
  | 'failed'
  | 'partial'
  | 'partial_failed'
  | 'cancelled'
  | 'uninstalling'
  | 'uninstalled';

export interface InstallerDeploymentTarget {
  machineId: string;
  status: InstallerDeploymentTargetStatus;
  error?: string | null;
  cancelledAt?: string | null;
  retriedAt?: string | null;
  [key: string]: unknown;
}

export interface InstallerDeploymentSummary {
  id: string;
  name: string;
  installer_name: string;
  installer_url: string;
  silent_flags: string;
  verify_path: string | null;
  sha256_checksum: string | null;
  parallel_install: boolean;
  targets: InstallerDeploymentTarget[];
  status: InstallerDeploymentStatus;
  createdAt: string | null;
  completedAt: string | null;
  updatedAt: string | null;
}

export interface InstallerDeploymentDetail extends InstallerDeploymentSummary {
  siteId: string;
}

export interface ListInstallerDeploymentsOptions {
  pageSize?: number;
  pageToken?: string;
}

export interface ListInstallerDeploymentsResult {
  items: InstallerDeploymentSummary[];
  nextPageToken: string;
}

export interface CreateInstallerDeploymentOptions {
  name: string;
  installer_url: string;
  installer_name: string;
  silent_flags: string;
  machines: readonly string[];
  verify_path?: string;
  sha256_checksum?: string;
  close_processes?: readonly string[];
  suppress_projects?: readonly string[];
  parallel_install?: boolean;
  idempotencyKey?: string;
}

export interface CreateInstallerDeploymentResult {
  deploymentId: string;
  siteId: string;
  status: InstallerDeploymentStatus;
  targets: InstallerDeploymentTarget[];
}

export interface InstallerDeploymentMutationResult {
  deploymentId: string;
  siteId: string;
  status: InstallerDeploymentStatus;
  /** Number of targets affected by the mutation (retried/cancelled/queued). */
  retried?: number;
  cancelled?: number;
  queued?: number;
  machine_ids?: string[];
}

export interface InstallerDeploymentDeleteResult {
  deploymentId: string;
  siteId: string;
  deleted: boolean;
}

/* --------------------------------------------------------------------- */
/*  resource                                                             */
/* --------------------------------------------------------------------- */

export class InstallerDeployments {
  constructor(private readonly client: RoostClient) {}

  async list(
    siteId: string,
    opts: ListInstallerDeploymentsOptions = {},
  ): Promise<ListInstallerDeploymentsResult> {
    const res = await this.client.request<{
      items: InstallerDeploymentSummary[];
      next_page_token: string;
    }>(`/api/sites/${encodeURIComponent(siteId)}/deployments`, {
      query: {
        page_size: opts.pageSize,
        page_token: opts.pageToken,
      },
    });
    return {
      items: res.data.items,
      nextPageToken: res.data.next_page_token,
    };
  }

  async get(siteId: string, deploymentId: string): Promise<InstallerDeploymentDetail> {
    const res = await this.client.request<InstallerDeploymentDetail>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments/${encodeURIComponent(deploymentId)}`,
    );
    return res.data;
  }

  async create(
    siteId: string,
    opts: CreateInstallerDeploymentOptions,
  ): Promise<CreateInstallerDeploymentResult> {
    const body: Record<string, unknown> = {
      name: opts.name,
      installer_url: opts.installer_url,
      installer_name: opts.installer_name,
      silent_flags: opts.silent_flags,
      machines: [...opts.machines],
    };
    if (opts.verify_path !== undefined) body.verify_path = opts.verify_path;
    if (opts.sha256_checksum !== undefined) body.sha256_checksum = opts.sha256_checksum;
    if (opts.close_processes !== undefined) body.close_processes = [...opts.close_processes];
    if (opts.suppress_projects !== undefined) body.suppress_projects = [...opts.suppress_projects];
    if (opts.parallel_install) body.parallel_install = true;

    const res = await this.client.request<CreateInstallerDeploymentResult>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments`,
      {
        method: 'POST',
        body,
        idempotencyKey:
          opts.idempotencyKey ?? `sdk-installer-deployments-create-${randomUUID()}`,
      },
    );
    return res.data;
  }

  async retry(
    siteId: string,
    deploymentId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<InstallerDeploymentMutationResult> {
    const res = await this.client.request<InstallerDeploymentMutationResult>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments/${encodeURIComponent(deploymentId)}/retry`,
      {
        method: 'POST',
        body: {},
        idempotencyKey:
          opts.idempotencyKey ?? `sdk-installer-deployments-retry-${randomUUID()}`,
      },
    );
    return res.data;
  }

  async cancel(
    siteId: string,
    deploymentId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<InstallerDeploymentMutationResult> {
    const res = await this.client.request<InstallerDeploymentMutationResult>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments/${encodeURIComponent(deploymentId)}/cancel`,
      {
        method: 'POST',
        body: {},
        idempotencyKey:
          opts.idempotencyKey ?? `sdk-installer-deployments-cancel-${randomUUID()}`,
      },
    );
    return res.data;
  }

  async uninstall(
    siteId: string,
    deploymentId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<InstallerDeploymentMutationResult> {
    const res = await this.client.request<InstallerDeploymentMutationResult>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments/${encodeURIComponent(deploymentId)}/uninstall`,
      {
        method: 'POST',
        body: {},
        idempotencyKey:
          opts.idempotencyKey ?? `sdk-installer-deployments-uninstall-${randomUUID()}`,
      },
    );
    return res.data;
  }

  async delete(
    siteId: string,
    deploymentId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<InstallerDeploymentDeleteResult> {
    const res = await this.client.request<InstallerDeploymentDeleteResult>(
      `/api/sites/${encodeURIComponent(siteId)}/deployments/${encodeURIComponent(deploymentId)}`,
      {
        method: 'DELETE',
        body: {},
        idempotencyKey:
          opts.idempotencyKey ?? `sdk-installer-deployments-delete-${randomUUID()}`,
      },
    );
    return res.data;
  }
}
